### PR TITLE
Authentication via Telegram

### LINKED ISSUE #14: Setup Authentication
The plan is to use telegram to authenticate user.  

The login workflow is as follows:

1. Telegram authentication using telegram login widget (github pages in dev)
2. Send authentication details to backend /login route
3. Check authenticity of data
4. Check if tele_id exists and populate table if handle exists
5. Generate session token

### PR DESCRIPTION
**Pull request by zupey**
This PR sets up the authentication aspect of the booking system and closes #14.

It accomplishes the following milestones:
- [x] Checks the authenticity of the telegram payload sent by the frontend
- [x] Sends the session token back to the client

It does not:
* check the database and populate the relevant tables

## How to test
1. Add a `BOT_TOKEN` field in your `.env` file.  Contact @misterpuffin for the token.
2. Go to the frontend website's login page (currently https://usdevs.github.io/uscwebsite-hackathon-backend/)
3. Login via telegram
4. The token should be displayed in the top box
image
5. Access the main route `localhost:3000/` with an extra `Authentication` header in the http `GET` request with the token as the value.
How I did it is by sending the request via developer tools in Firefox
image
6. The server should respond with response `200`.

### CODE REVIEW

File: src/routes/routes.ts [hunk 1]
```diff
@@ -1,23 +1,25 @@
...
+import bodyParser from 'body-parser'
+import { getBookings } from '../controllers/bookings'
+import { editBooking } from '../controllers/edit'
+import { deleteBooking } from '../controllers/delete'
 
 export const router: Router = Router()
 
 // landing page
-router.get('/', (req: Request, res: Response) => {
+router.get('/', authenticate, (req: Request, res: Response) => {
```
Comment (SeekSaveServe) [inline src/routes/routes.ts:13]: actual route for landing page shouldn't require auth

**Review by misterpuffin — APPROVED** [review]
lgtm! thanks for the changes. i think in the future we could perhaps move the crypto related functions to a utils folder, and move the interfaces like TelegramAuth to its own types folder.

also not relevant to this PR, but I think the imports can be made neater by changing our folder structure. 
Instead of: 
```import { getBookings } from '../controllers/bookings'
import { editBooking } from '../controllers/edit'
import { deleteBooking } from '../controllers/delete'
```

Perhaps this would be neater.
``` import { getBookings, editBooking, deleteBooking } from '../controllers/bookings' ```

Would also prefer absolute file imports, but just a personal preference of mine.

File: src/routes/routes.ts (hunk 1 above)
Comment (zupey) [reply src/routes/routes.ts]: Ok, updated routes for those that should be public

**Comment by zupey** [comment]
Thanks for your feedback! perhaps we can refactor the file structure etc. in a separate PR

On another note, I just tested whether the `jsonwebtoken` library is subject to the `none-algorithm` attack.  Fortunately, the library ensures that the digital signature is a required field.  This is the tutorial I followed to launch the `none-algorithm` attack: https://blog.pentesteracademy.com/hacking-jwt-tokens-the-none-algorithm-67c14bb15771
